tests/runtime/test_dbk_onnx_inference.c

```diff
@@ -0,0 +1,218 @@
+  cl_tensor_desc initializer_tensor_descs[] = { { 1,
+                                                  CL_TENSOR_DTYPE_FP32,
+                                                  { CL_TENSOR_PROPERTY_NONE },
+                                                  { 1 },
+                                                  &tensor_layout,
+                                                  CL_TENSOR_LAYOUT_ML } };
+
+  const cl_dbk_attributes_khr_onnx_inference onnx_inference_attributes
+    = { model_bytes, model_size, 2, input_tensor_names, input_tensor_descs, 1,
+        output_tensor_names, output_tensor_descs,
+        /* The below attributes are optional and can be left zeroed out */
+        1, initializer_names, initializer_tensor_descs,
+        (const char **)initializer_data };
+  const void *dbk_attributes[] = { &onnx_inference_attributes };
+  cl_int device_support = 0;
+  cl_program program = ext.clCreateProgramWithDefinedBuiltInKernels (
+    context, 1, &device, 1, &dbk_id, &dbk_name, dbk_attributes,
+    &device_support, &error);
+  CHECK_CL_ERROR (error);
+  free(model_bytes);
+  CHECK_CL_ERROR (clBuildProgram (program, 1, &device, NULL, NULL, NULL));
+  cl_kernel kernel = clCreateKernel (program, "exp_onnx_inference", &error);
+  CHECK_CL_ERROR (error);
+
+  cl_mem_properties mem_props[]
+    = { CL_MEM_TENSOR, (cl_mem_properties)&tensor_desc, 0 };
+
+  float *test_data = NULL;
+  MALLOC_CHECKED (test_data, sizeof (float) * num_elements * 2);
+  float *in_A = test_data;
+  float *in_B = test_data + num_elements;
+
+  srand (clock ());
+  for (size_t i = 0; i < num_elements; ++i)
+    {
+      in_A[i] = (float)rand () / (float)RAND_MAX;
+      in_B[i] = (float)rand () / (float)RAND_MAX;
+    }
+
+  cl_mem all_inputs = clCreateBufferWithProperties (
+    context, mem_props, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
+    sizeof (float) * num_elements * 2, test_data, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem all_outputs = clCreateBufferWithProperties (
+    context, mem_props, CL_MEM_WRITE_ONLY, sizeof (float) * num_elements, NULL,
+    &error);
+  CHECK_CL_ERROR (error);
+  uint64_t input_offsets[2] = { 0, num_elements * sizeof (float) };
+  uint64_t output_offsets[1] = { 0 };
+  cl_mem input_offsets_buf
+    = clCreateBuffer (context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
+                      sizeof (uint64_t) * 2, input_offsets, &error);
+  CHECK_CL_ERROR (error);
+  cl_mem output_offsets_buf
+    = clCreateBuffer (context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
+                      sizeof (uint64_t) * 1, output_offsets, &error);
+  CHECK_CL_ERROR (error);
+
+  CHECK_CL_ERROR (
+    clSetKernelArg (kernel, 0, sizeof (cl_mem), &input_offsets_buf));
+  CHECK_CL_ERROR (clSetKernelArg (kernel, 1, sizeof (cl_mem), &all_inputs));
+  CHECK_CL_ERROR (
+    clSetKernelArg (kernel, 2, sizeof (cl_mem), &output_offsets_buf));
+  CHECK_CL_ERROR (clSetKernelArg (kernel, 3, sizeof (cl_mem), &all_outputs));
```

Comment:
Since ONNX model takes inputs and outputs as tensors, shouldn't the ONNX kernel be interfaced with tensors too?
Can't we have the `clCreateProgramWithDefinedBuiltInKernels` to create a kernel with N + M kernel (tensor) parameters based on N inputs and M outputs in the DBK attribute structure?

Reply:
That was my original idea, but the conclusion on that was that that would have too much of a risk of running into unfeasible numbers of kernel arguments. Trivially so, if we also want to override initializers at that level.

Reply:
Is the concern here that the `CL_DEVICE_MAX_PARAMETER_SIZE` could be exceeded on some models?

I agree there potentially could be unfeasible number of kernel arguments if the initializers would be exposed as arguments too but I'm not sure about that with the I/O tensors alone as kernel arguments. Are there such models that have more than ten or even hundred I/O tensors?

Reply:
No idea, I haven't looked at a lot of models. DeepFilterNet consists of 3 networks with the most verbose one having 9 inputs/outputs. I don't think `CL_DEVICE_MAX_PARAMETER_SIZE` would be the problem since the parameters are just `cl_mem` in the end.